packages/neovim/src/utils/transport.ts

```diff
@@ -42,86 +43,89 @@ class Transport extends EventEmitter {
 
   private nextRequestId = 1;
 
-  private encodeStream: any;
-
-  private decodeStream: any;
-
   private reader: NodeJS.ReadableStream;
 
   private writer: NodeJS.WritableStream;
 
-  protected codec: msgpack.Codec;
+  private readonly extensionCodec: ExtensionCodec = this.initializeExtensionCodec();
 
   // Neovim client that holds state
   private client: any;
 
-  constructor() {
-    super();
-
-    const codec = this.setupCodec();
-    this.encodeStream = msgpack.createEncodeStream({ codec });
-    this.decodeStream = msgpack.createDecodeStream({ codec });
-    this.decodeStream.on('data', (msg: any[]) => {
-      this.parseMessage(msg);
-    });
-    this.decodeStream.on('end', () => {
-      this.detach();
-      this.emit('detach');
-    });
-  }
-
-  setupCodec() {
-    const codec = msgpack.createCodec();
+  private initializeExtensionCodec() {
+    const codec = new ExtensionCodec();
 
     Metadata.forEach(({ constructor }, id: number): void => {
-      codec.addExtPacker(id, constructor, (obj: any) =>
-        msgpack.encode(obj.data)
-      );
-      codec.addExtUnpacker(
-        id,
-        data =>
-          new constructor({
+      codec.register({
+        type: id,
+        encode: (input: any) => {
+          if (input instanceof constructor) {
+            return encode(input.data);
+          }
+          return null;
+        },
+        decode: data => {
+          return new constructor({
             transport: this,
             client: this.client,
-            data: msgpack.decode(data),
-          })
-      );
+            data: decode(data),
+          });
+        },
+      });
     });
 
-    this.codec = codec;
-    return this.codec;
+    return codec;
+  }
+
+  private encodeToBuffer(value: unknown) {
+    const encoded = encode(value, { extensionCodec: this.extensionCodec });
+    return Buffer.from(encoded.buffer, encoded.byteOffset, encoded.byteLength);
   }
 
   attach(
     writer: NodeJS.WritableStream,
     reader: NodeJS.ReadableStream,
     client: any
   ) {
-    this.encodeStream = this.encodeStream.pipe(writer);
-    const buffered = new Buffered();
-    reader.pipe(buffered).pipe(this.decodeStream);
     this.writer = writer;
     this.reader = reader;
     this.client = client;
-  }
 
-  detach() {
-    this.encodeStream.unpipe(this.writer);
-    this.reader.unpipe(this.decodeStream);
+    this.reader.on('end', () => {
+      this.emit('detach');
+    });
+
+    const asyncDecodeGenerator = decodeStream(this.reader as any, {
+      extensionCodec: this.extensionCodec,
+    });
+
+    // naively iterate async generator created via decodeStream.
+    // when runtime / polyfill allows replace to `for await (const val of asyncDecodeGenerator)`
+    // syntax instead.
```

Comment:
I believe TypeScript supports `for await` statement. For example, I could compile following code with TypeScript 3.7:

```typescript
function sleep(i: number) {
    return new Promise(resolve => setTimeout(resolve, i));
}

async function* f() {
  yield 1;
  await sleep(1000);
  yield 2;
}

async function main() {
    for await (const x of f()) {
        console.log(x);
    }
}

main().catch(console.error);
```

Reply:
It doesb but current eslint setup says it's not allowed due to concern of runtime polyfill on node does not support native async.

Reply:
I see. `"target": "es2015"` might be too old for translating `for await` statement (it introduces many runtime code).